Zend/zend_property_hooks.c

```diff
@@ -0,0 +1,215 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Ilija Tovilo <ilutov@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_property_hooks.h"
+
+typedef struct {
+	zend_object_iterator it;
+	bool by_ref;
+	zval properties;
+	zval current_data;
+} zend_hooked_object_iterator;
+
+static zend_result zend_hooked_object_it_valid(zend_object_iterator *iter);
+
+ZEND_API zend_array *zend_hooked_object_build_properties(zend_object *zobj)
+{
+	zend_class_entry *ce = zobj->ce;
+	zend_array *properties = zend_new_array(ce->default_properties_count);
+	zend_hash_real_init_mixed(properties);
+
+	zend_property_info *prop_info;
+	int backed_property_count = 0;
+	ZEND_HASH_MAP_FOREACH_PTR(&ce->properties_info, prop_info) {
+		if (prop_info->flags & ZEND_ACC_STATIC) {
+			continue;
+		}
+		if (!(prop_info->flags & ZEND_ACC_VIRTUAL)) {
+			backed_property_count++;
+		}
+		if (prop_info->hooks) {
+			_zend_hash_append_ptr(properties, prop_info->name, prop_info);
+		} else {
+			if (UNEXPECTED(Z_TYPE_P(OBJ_PROP(zobj, prop_info->offset)) == IS_UNDEF)) {
+				HT_FLAGS(properties) |= HASH_FLAG_HAS_EMPTY_IND;
+			}
+			_zend_hash_append_ind(properties, prop_info->name, OBJ_PROP(zobj, prop_info->offset));
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	if (zobj->properties) {
+		zend_string *prop_name;
+		zval *prop_value;
+		ZEND_HASH_FOREACH_STR_KEY_VAL_FROM(zobj->properties, prop_name, prop_value, backed_property_count) {
+			Z_TRY_ADDREF_P(_zend_hash_append(properties, prop_name, prop_value));
+		} ZEND_HASH_FOREACH_END();
+	}
```

Comment:
This fetching of properties ahead of time leads to a subtle behaviour difference between classes with hooked properties and classes without:
```
# ./sapi/cli/php -r 'class A extends stdClass { public $foo; } $a = new A; foreach ($a as $k => &$v) { if ($k == "foo") $a->bar = "baz"; var_dump($k); }'
string(3) "foo"
string(3) "bar"
# ./sapi/cli/php -r 'class A extends stdClass { public $foo { &get => $this->foo; } } $a = new A; foreach ($a as $k => &$v) { if ($k == "foo") $a->bar = "baz"; var_dump($k); }'
string(3) "foo"
```

Reply:
I pushed a new commit to handle dynamic props better. https://github.com/php/php-src/pull/13455/commits/befd85b92d17c43a9d14a3c1661abcb940c24a42 It will probably require some more testing.